CSLabs.Api/Controllers/ModuleController.cs

```diff
@@ -27,11 +28,27 @@ public class ModuleController : BaseController
         public ModuleController(BaseControllerDependencies deps) : base(deps)
         {
         }
-        // GET api/values
+        // GET api/values?tagNames
         [HttpGet]
-        public IActionResult Get()
+        public IActionResult Get(string tagNames = "")
         {
-            return Ok(DatabaseContext.Modules.Where(m => m.Published).ToList());
+            if (string.IsNullOrEmpty(tagNames))
+            {
+                return Ok(DatabaseContext.Modules.Where(m => m.Published).ToList());
+            }
+            string[] tags = tagNames.Split(",");
+            var moduleTags = new List<ModuleTag>();
+            foreach (string tagName in tags)
+            {
+                moduleTags = new List<ModuleTag>(DatabaseContext.Tags
+                    .Where(t => t.Name.Equals(tagName))
+                    .SelectMany(t => t.ModuleTags).ToList());
+            }
+            var modules = moduleTags
+                .Select(mt => this.DatabaseContext.Modules.Single(m => m.Id == mt.ModuleId))
+                    .ToList()
+                    .Distinct();
+            return Ok(modules);
```

Comment:
This method can be reduced to a few lines.
Structure your queries so that the table  you want to end up with is the starting data set.
If you want modules then write `DatabaseContext.Modules` and figure out how to filter them down from there.

```suggestion
        // GET api/v1/Modules/?tagNames
        [HttpGet]
        public async Task<IActionResult> Get(string tagNames = "")
        {
            if (string.IsNullOrEmpty(tagNames))
                return Ok(DatabaseContext.Modules.Where(m => m.Published).ToList());
            var tagNamesList = tagNames.Split(",").Select(s => s.Trim()).ToList();
            return Ok(
                await DatabaseContext.Modules
                .Where(m => m.ModuleTags.Any(mt => tagNamesList.Contains(mt.Tag.Name)))
                .ToListAsync()
            );
        }
```